ffi/framebuffer_linux.lua

```diff
@@ -13,228 +13,124 @@ local framebuffer = {
-    if self.device:has3BytesWideFrameBuffer() then
-        vinfo.bits_per_pixel = 24
-        vinfo.xres = vinfo.xres / 3
-    end
+    -- Get screen information
+    assert(C.ioctl(self.fd, C.FBIOGET_FSCREENINFO, finfo) == 0, "cannot get fixed screen info")
+    assert(C.ioctl(self.fd, C.FBIOGET_VSCREENINFO, vinfo) == 0, "cannot get variable screen info")
 
-    -- Recap final, potentially tweaked FB setup...
-    io.write("FB: computed size: ", self.fb_size, "\n")
-    io.write("FB: id           : ", ffi.string(finfo.id), "\n")
-    io.write("FB: smem_len     : ", finfo.smem_len, "\n")
-    io.write("FB: line_length  : ", finfo.line_length, "\n")
-    io.write("FB: xres         : ", vinfo.xres, "\n")
-    io.write("FB: xres_virtual : ", vinfo.xres_virtual, "\n")
-    io.write("FB: yres         : ", vinfo.yres, "\n")
-    io.write("FB: yres_virtual : ", vinfo.yres_virtual, "\n")
-    io.write("FB: bpp          : ", vinfo.bits_per_pixel, "\n")
-    io.flush()
+    -- Apply frontend kludges (color lux, very old eink...)
+    self:fbinfoOverride(finfo, vinfo)
+
+    assert(finfo.type == C.FB_TYPE_PACKED_PIXELS, "video type not supported")
+    assert(vinfo.xres > 0 and vinfo.yres > 0, "invalid framebuffer resolution")
+
+    -- We can forgo all messy fb detection logic by simply assuming only f.line_length and v.yres are valid,
+    -- because if they aren't, the world would be on fire no matter what sooner or later.
+    self.fb_size = finfo.line_length * vinfo.yres
+
+    local bpp = vinfo.bits_per_pixel
+    local stride_pixels = finfo.line_length * 8
+    assert(stride_pixels % bpp == 0, "line_length doesn't end at pixel boundary")
+    stride_pixels = stride_pixels / bpp
+
+    self.debug("FB info (post fixup)", {
+          fb_size = self.fb_size,
+          xres = vinfo.xres,
+          yres = vinfo.yres,
+          xoffset = vinfo.xoffset,
+          yoffset = vinfo.yoffset,
+          bpp = bpp,
+          xres_virtual = vinfo.xres,
+          yres_virtual = vinfo.yres,
+          line_length = finfo.line_length,
+          stride_pixels = stride_pixels,
+          smem_len = finfo.smem_len,
+          type = finfo.type,
+          mmio_len = finfo.mmio_len,
+          rotate = vinfo.rotate,
+          width_mm = vinfo.width,
+          height_mm = vinfo.height,
+    })
+
+    assert(vinfo.xres > 0 and vinfo.yres > 0, "invalid framebuffer resolution")
 
     -- Make sure we never try to map a larger memory region than the fb reports
-    assert(self.fb_size <= finfo.smem_len, "computed fb memory region too large")
+    -- @warning Feel free to remove this check if it burns. There are chinese things out there that even happily report smem as 0x1000 and such.
+    assert(self.fb_size <= finfo.smem_len or finfo.smem <= 0x1000, "computed fb memory region too large")
 
     self.data = C.mmap(nil,
-                           self.fb_size,
+                           bit.band(self.fb_size+4095,-4096),
```

Comment:
I wouldn't mind a little comment about the alignment trickery here for future readers ;).